### DISCUSSION TITLE
Are there custom step integrations for Allure Python?

### DISCUSSION THREAD

**Discussion by hwentzke** [allure-framework/allure2#1708] [Questions & Support] [answered]
Hello,

I have been working with Allure for a long time in Java. Since 2021 I am working with it in Python and I am missing a very nice feature that exists for Allure in Java.

The feature is to create custom steps in Java
https://github.com/allure-framework/allure-java/pull/282

Question is: is there something like this possible for Allure Python? I have been trying a bit, but couldn't make it work, and also couldn't find anything online.
The fact that this feature is hidden in an issue on github for Java too is not great either, this should be in the documentation, as it is really helpful to improve the structure of steps in your reports!

**Comment by skhomuti** [comment, marked as answer]
Hi @hwentzke! 

Let's get some details
Which framework do you use? If pytest, you can take a look some examples here https://github.com/allure-examples/allure-examples/blob/master/allure-python/allure-python-pytest/test/allure_step_test.py

I agree that there are no some features in python, let's decide what exactly you are missing and I will help with

As I know at least pytest doesn't have log-style steps at the moment, like
```
step("set up test database", Status.SKIPPED);
```
but probably I missing something else

**Reply by hwentzke** [reply]
Hi,

I am using Pytest for Python.
And I looked at the code examples for Steps, but I don't see the feature I would love to use.

In Java, I use the Custom Lambda steps to create my own groups of attachments without using the @Step annotations.
Because the annotated steps are attaching too much baggage into the reports that I don't want, e.g. all the parameters given to a function that is annotated as Step.
In java, I have a function that has several arguments to be attached to the reports, and then I call this function in my test execution when I have all the required attachment data available.

```
    public static void addAllureStepWithAttachments(String stepName,
                                                    String name_requestId, String requestId,
                                                    String name_instanaId, String instanaId,
                                                    String name_status, String statuscode,
                                                    String name_repsonseBody, String responseBody,
                                                    String name_curl, String curl) {
        step(stepName, (step) -> {
            attachment(name_requestId, requestId);
            attachment(name_instanaId, instanaId);
            attachment(name_status, statuscode);
            attachment(name_repsonseBody, responseBody);
            attachment(name_curl, curl);
        });
    }
```

This is what I use for Java.

**Reply by skhomuti** [reply]
Take a look this python snippet
Custom python context managers similar to your function
```python
from contextlib import contextmanager


@contextmanager
def step_with_attachments(step_name, **kwargs):
    with allure.step(step_name):
        for name, param in kwargs.items():
            allure.attach(param, name=name)
        yield


def test_with_custom_step():
    with step_with_attachments("Some step", param1="first", param2="second"):
        do_something_else()
```

or you just can make a function if you want
```python
def add_step_with_attachments(step_name, **kwargs):
    with allure.step(step_name):
        for name, param in kwargs.items():
            allure.attach(param, name=name)


def test_with_custom_step():
    add_step_with_attachments("Some step", param1="first", param2="second")
    do_something_else()
```

does it solve your question? We can add this example to `allure-examples` repo

**Reply by hwentzke** [reply]
I have implemented the 2nd solution and it works exactly as I want it.
I get my Step with assigned name and no parameter attachments and then attach my data to the step.

Thank you for you help!
